File: kai_universe_backend/app/modules/downloads/domain/repositories/download_repository.py
from abc import ABC, abstractmethod
from app.modules.downloads.domain.entities.download import Download


class DownloadRepository(ABC):

    @abstractmethod
    def list(self) -> list[Download]:
        pass

    @abstractmethod
    def find_by_id(self, download_id: str,) -> Download | None:
        pass

    @abstractmethod
    def enqueue(self, model_id: str, quantization: str, total_bytes: int,) -> Download:
        pass
    
    @abstractmethod
    def update_status(
        self,
        download_id: str,
        status: str,
    ) -> None:
        pass
    
    @abstractmethod
    def mark_completed(
        self,
        download_id: str,
    ) -> None:
        pass
    
    @abstractmethod
    def mark_failed(
        self,
        download_id: str,
        error_message: str,
    ) -> None:
        pass

    @abstractmethod
    def pause(self, download_id: str,) -> None:
        pass

    @abstractmethod
    def resume(self, download_id: str,) -> None:
        pass

    @abstractmethod
    def cancel(self, download_id: str,) -> None:
        pass

File: kai_universe_ui/src/shared/domain/download/ports/DownloadRepository.ts
import type { ModelId } from '../../model/value-objects/ModelId';
import type { Quantization } from '../../model/value-objects/Quantization';
import type { Bytes } from '../../primitives/Bytes';
import type { Download, DownloadId, DownloadProgress } from '../entities/Download';

export interface EnqueueDownloadInput {
    readonly modelId: ModelId;

    readonly hfRepository: string;

    readonly quantization: Quantization;

    readonly sizeBytes?: Bytes;
}

export interface DownloadRepository {
  list(): Promise<readonly Download[]>;
  findById(id: DownloadId): Promise<Download | null>;
  enqueue(input: EnqueueDownloadInput): Promise<Download>;
  pause(id: DownloadId): Promise<void>;
  resume(id: DownloadId): Promise<void>;
  cancel(id: DownloadId): Promise<void>;

  // Hot updates from the simulator. Adapter implements either polling or push.
  observe(id: DownloadId): AsyncIterable<DownloadProgress>;
}
